DRILL-6096: Provide mechanism to configure text writer configuration

arina-ielchiieva opened the pull request:
1. Usage of format plugin configuration allows to specify line and field delimiters, quotes and escape characters.
2. Usage of system / session options allows to specify if writer should add headers, force quotes.

Jira - [DRILL-6096](https://issues.apache.org/jira/browse/DRILL-6096).

arina-ielchiieva commented:
@vvysotskyi please review.

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/text/writer/TextRecordWriter.java:
```diff
@@ -93,29 +112,35 @@ public void startNewSchema(List<String> columnNames) throws IOException {
+
+    if (writerSettings.isHeaderWritingEnabled()) {
+      writer.writeHeaders(
+        IntStream.range(0, schema.getFieldCount())
```
Nit: it can be a little bit optimized to avoid looking up elements from the list by index:
```suggestion
        StreamSupport.stream(schema.spliterator(), false)
```

arina-ielchiieva replied:
Good point, thanks for noticing. Fixed.

paul-rogers commented:
General question: the idea of giving the user more control is a good one. I'm seeing this need in multiple of recent plugins. I wonder, however, if we should offer a more scalable solution?

We currently have plugin, session, config and table function options all which control some aspect of reading and writing. The plugin and config options really want to be "bulk" or "slow moving" settings: set them once for a system or file format.

Session options have two problems. First, they introduce client state: they must be set and reset around each query that needs them. This prevents load balancing and fail-over. (By contrast, Impala keeps options in the client and sends them to the server on each query, allowing unlimited load balancing via round-robin and allows transparent fail over.)

Further, session options are awkward for many tools: in addition to running a query, the client must send `ALTER SESSION` statements which are somehow associated with that query. Easy for our unit tests, not so easy for most BI tools.

Table options look like what we want, but they have a number of issues, including that, when used, they "erase" any options set in the format plugin.

Is there some other solution we could consider? Better table options? Some syntax that allows setting of options in the query? Some kind of view or schema that encapsulates these options so that they don't have to be provided on every query?

I don't have a solid suggestion; I'm simply pointing out the ease-of-use issue.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/text/writer/TextRecordWriter.java:
```diff
@@ -165,10 +182,12 @@ public void writeField() throws IOException {
+
+    fRecordStarted = false;
+    if (writer != null) {
+      writer.close();
```
This is a writer. The close can fail with an exception. If it does, the entire operation should fail. Should we throw a UserException` with the details here rather than let the `IOException` bubble up?`

arina-ielchiieva replied:
Good point, will update the code.

paul-rogers commented on exec/java-exec/src/test/java/org/apache/drill/test/ClusterFixture.java:
```diff
@@ -57,10 +40,27 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
```
Can we define a Drill standard for import order? The Apache settings I imported into Eclipse are `java`, `javax`, `org`, `com`. Statics are at the top.

In this change, looks like `java` moves the to bottom.

arina-ielchiieva replied:
I use standard IntelliJ IDEA imports order:

<img width="1331" alt="image" src="https://user-images.githubusercontent.com/15086720/67193155-811dcc80-f3fd-11e9-8434-53b9a5f598cf.png">

Regarding updating the check styles, there is a Jira (https://issues.apache.org/jira/browse/DRILL-7352) where we can post comments.

arina-ielchiieva commented:
@paul-rogers thanks for the code review, addressed code review comments, force-pushed since there were minor changes in the code. 

Regarding design, the aim of this Jira was just to fix text writer to write proper text files: before if column contained field separator, field was not enclosed in the quotes, thus we were writing text files which Drill could not read. Now when user indicates write format using session option (this is common approach for all formats), Drill produces text files, it can read back. Basically, if user has configured format plugin:
```
  "formats": {
    "csvh": {
      "type": "text",
      "extensions": [
        "csvh"
      ],
      "lineDelimiter": "\n",
      "fieldDelimiter": ",",
      "extractHeader": true
    }
   },
```
Drill will be able to read and write such text files correctly. Same approach is used for `parquet`, `json`. All user needs to do is to indicate write format using session option: `alter session set `store.format` = 'csvh';` (`parquet`, `json`). I am not saying this is ideal and we might need to reconsider such writing approach but I guess not in the scope of Jira since such re-design would touch all file writers.

paul-rogers commented:
To answer my other comment; about the lack of ease-of-use with the current session options related to file formats: the answer is the schema mechanism you defined a while back. Even if no schema is provided, we should allow the user to override formatting properties using the table options encoded in that schema.

Then, I can write a file using, say pipe delimiters, have that be recorded, and read the file automatically using those delimiters. Probably work to be done, but would we a nice solution that does not require HMS.

arina-ielchiieva commented:
Sounds interesting, basically, you are suggesting to create implicit schema file when writing a text files (i.e. creating table), thus format options will be picked up directly from schema file when reading table back.
